server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -116,11 +127,13 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
         }
 
         checkIndexBlocks(clusterState, concreteIndices);
-
-        final Map<String, FieldCapabilitiesIndexResponse> indexResponses = Collections.synchronizedMap(new HashMap<>());
         // This map is used to share the index response for indices which have the same index mapping hash to reduce the memory usage.
         final Map<String, FieldCapabilitiesIndexResponse> indexMappingHashToResponses = Collections.synchronizedMap(new HashMap<>());
         final Consumer<FieldCapabilitiesIndexResponse> handleIndexResponse = resp -> {
+            if (fieldCapTask.notifyIfCancelled(listener)) {
+                LOGGER.trace("stop accumulating index responses on cancelled");
+                return;
```

Comment:
I'd prefer to do this in a way that avoids the race which might lead to retaining an index response even after cancellation (i.e. if the cancellation lands just after this check). See for instance `TransportBroadcastByNodeAction#executeAsCoordinatingNode` for a neat-ish way to do that. But I'm not going to insist if @romseygeek thinks this is ok.

Reply:
I will defer to @DaveCTurner on this!

Reply:
ok, I've reworked this using RefCountingRunnable. See: https://github.com/elastic/elasticsearch/pull/93716/commits/c60f98b09522a75111c6540db55cf32f00864c3a